### PR TITLE
Fused L2 1-NN based on cutlass 3xTF32 / DMMA

### PR DESCRIPTION
**Pull request by mdoijade**
-- 3xTF32 & DMMA cutlass based persistent FusedL2NN kernel version loosely based on grouped gemm but customized for single problem size. 
-- as the value of `k` increases the performance benefit of this implementation gets better. 
for k==64 upto 1.3x, for k ==128 upto 1.53x, k == 256, up to 1.67x.
-- for all the sizes of `k`  this kernel out performs previous implementation.
-- attaching the results of FusedL2NN Benchmark of previous implementation with this cutlass version.

### CODE REVIEW

**Comment by mdoijade** [comment]
+ @teju85 @tfeher  for visibility and review.

**Comment by ajschmidt8** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: cpp/include/raft/distance/detail/fused_l2_nn_gemm.h
Comment (tfeher) [inline cpp/include/raft/distance/detail/fused_l2_nn_gemm.h:73]: Just a note: right now you are setting it to the 3xTF32 variant so that we do not loose precision. In a follow up work, we would want to enable lower precision implementations as well, like TF32, or FP32 for internal calculation.

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Mahesh, thank you for this PR!

It is nice to see the performance improvement for large-k cases, but the perf loss for small k -large n combination should be avoided, as that case can occur for ANN index building.

We have discussed online that you are looking into ways to express the operation in cutlass with less atomic ops. I am looking forward to see that version. In the meantime I am sharing my comments for the current implementation.

**Comment by mdoijade** [comment]
@tfeher Thanks for the review, I will update these changes soon in this PR.
as discussed offline about the recent improvements made to this cutlass kernel, please find the results of DISTANCE_BENCH.
please check only `float` inputs results as I haven't touched the `double`  kernel.

Now for all k == 128 cases cutlass version is better with these improvements of new gemm shape + shared memory based coalesced gmem writes in the critical section.
[FusedL2NN.cutlass.vs.fp32.simt.kernel_22_feb_updated_float_results.xlsx](https://github.com/rapidsai/raft/files/10806122/FusedL2NN.cutlass.vs.fp32.simt.kernel_22_feb_updated_float_results.xlsx)

File: cpp/include/raft/distance/detail/fused_l2_nn_gemm.h
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_l2_nn_gemm.h:73]: sure, noted.

File: cpp/include/raft/distance/detail/predicated_tile_iterator_reduced_vec.h
Comment (mdoijade) [reply cpp/include/raft/distance/detail/predicated_tile_iterator_reduced_vec.h:28]: No, they are used for passing mutex array & reduction functor in the tile iterator

File: cpp/include/raft/distance/detail/fused_l2_nn_epilogue_elementwise.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_l2_nn_epilogue_elementwise.cuh:23]: actually this does not refer to reduction, it refers to converting the distance value (float/double) to key-value pair type.

File: cpp/include/raft/distance/detail/fused_l2_nn_cutlass_base.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_l2_nn_cutlass_base.cuh:160]: this would require me to additional pass raft handle in this function. should I do that?

Comment (cjnolet) [reply cpp/include/raft/distance/detail/fused_l2_nn_cutlass_base.cuh:151]: @mdoijade you could pass the raft handle (now called raft::resources object, or raft::device_resources) or the rmm memory resource (from the raft::resources). Either one would be fine. We are trying to eliminate the explicit workspace allocations from user-facing APIs and instead allow the user to control (indirectly) how the workspaces are allocated.

File: cpp/include/raft/distance/detail/fused_l2_nn.cuh
Comment (cjnolet) [reply cpp/include/raft/distance/detail/fused_l2_nn.cuh:362]: If an existing operator can be used, I’d propose that it be used so that we can consolidate these types of building blocks wherever possible.

**Review by cjnolet — CHANGES_REQUESTED** [review]
I just realized I had some pending review comments sitting around for a little while. Sorry about that, Mahesh! 

Mostly minor things on my part as Tamas has done a more thorough review of the CUDA bits.

I also want to tag @ahendriksen for his thoughts / review as he recently refactored some of the distance dispatch details.

File: cpp/include/raft/distance/detail/fused_l2_nn.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/detail/fused_l2_nn.cuh:362]: I think this would work:
```
#include <raft/core/operators.hpp>
auto fin_op = raft::identity_op();
```

File: cpp/include/raft/distance/detail/fused_l2_nn.cuh
Comment (ahendriksen) [inline cpp/include/raft/distance/detail/fused_l2_nn.cuh:328]: PR #1142 moved things around a bit. The L2ExpandedOp can be obtained with:
```
#include <raft/distance/detail/distance_ops/l2_exp.cuh>

      using L2Op = ops::l2_exp_cutlass_op<DataT, AccT>;
      L2Op l2_op(perform_sqrt);
```
See: https://github.com/rapidsai/raft/blob/b12fc82be4c06cd157925ef14411fa2ee9cf0d07/cpp/include/raft/distance/detail/distance.cuh#L533

**Comment by mdoijade** [comment]
[FusedL2NN.cutlass.vs.fp32.simt.kernel_16_march_shfl_optimized_results.xlsx](https://github.com/rapidsai/raft/files/10993157/FusedL2NN.cutlass.vs.fp32.simt.kernel_16_march_shfl_optimized_results.xlsx)

**Comment by cjnolet** [comment]
@mdoijade just checking in on the status of this PR. Do you think these changes might be ready for 23.06 soon?

**Comment by mdoijade** [comment]
> @mdoijade just checking in on the status of this PR. Do you think these changes might be ready for 23.06 soon?

Yes it is almost done, I will post the final changes in this PR by end of this week.

File: cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh
Comment (tfeher) [inline cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh:80]: The corresponding file for the pairwise distance cutlass kernels sets this to `false` [link](https://github.com/rapidsai/raft/blob/576b22fccdd19f430c7f02bd29b4bc69dd2a0bf7/cpp/include/raft/distance/detail/pairwise_distance_epilogue_elementwise.h#L76). IIUC, the  `kIsHeavy = true` leads to the epilogue funtor not inlined. Is this beneficial for occupancy?

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks @mdoijade for the updates! The PR has changed significantly since the previous review round to enable an impressive speedup. Here are the first batch of comments of my second review.

This PR adds customized cutlass headers to implement the fused L2 NN operation.

While the changeset seem to be large, it is actually much smaller, if we compare the new cutlass headers to their original version (based on the reference added to files, and after applying raft formatting to the originals). Comparing that way makes it easier to follow how the cutlass code was adapted to our needs, and reveals a clean implementation. Great work, I have only smaller comments (so far)!

File: cpp/include/raft/distance/detail/fused_l2_nn.cuh
Comment (ahendriksen) [reply cpp/include/raft/distance/detail/fused_l2_nn.cuh:309]: Yes: using the fusedL2NNkernel pointer is a good choice.

File: cpp/include/raft/distance/detail/fused_distance_nn/predicated_tile_iterator_normvec_smem.h
Comment (tfeher) [inline cpp/include/raft/distance/detail/fused_distance_nn/predicated_tile_iterator_normvec_smem.h:157]: what is "di"?

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Mahesh for the updates so far. Here is my second batch of comments.

File: cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh:80]: yes need to set this as true here for occupancy reasons. without it there are register spills as we have a heavy reduction operation in epilogue.

File: cpp/include/raft/distance/detail/fused_distance_nn/predicated_tile_iterator_normvec_smem.h
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/predicated_tile_iterator_normvec_smem.h:157]: It is typo, fixed it now :)

File: cpp/include/raft/distance/detail/fused_distance_nn/epilogue.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/epilogue.cuh:50]: For this file it was 2.9

File: cpp/include/raft/distance/detail/fused_distance_nn/persistent_gemm.h
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/persistent_gemm.h:48]: done

File: cpp/include/raft/distance/detail/fused_distance_nn/cutlass_base.cuh
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/cutlass_base.cuh:146]: but then this will require me to pass the raft handle in this function. is it worth changing the `fusedL2NNImpl` API  for it?

**Comment by cjnolet** [comment]
I've lost track of where we are with this PR. Do you guys think this will make it into 23.06? (Burndown is in 3 days).

File: cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh
Comment (tfeher) [reply cpp/include/raft/distance/detail/fused_distance_nn/epilogue_elementwise.cuh:110]: Thanks for the explanation!

File: cpp/include/raft/distance/detail/fused_distance_nn/cutlass_base.cuh
Comment (tfeher) [reply cpp/include/raft/distance/detail/fused_distance_nn/cutlass_base.cuh:146]: Changing `fusedL2NNImpl` would not be a problem, since that is in `detail` namespace, but my suggestion would require changing the public `fusedL2NN` interface as well. Let's skip this for now.

File: cpp/include/raft/distance/detail/fused_distance_nn/persistent_gemm.h
Comment (tfeher) [reply cpp/include/raft/distance/detail/fused_distance_nn/persistent_gemm.h:48]: Thanks!

**Review by tfeher — COMMENTED** [review]
Thanks Mahesh for the updates!

@cjnolet, the PR is in a good shape, most of the issues have been addressed, it shall make it to 23.06.

**Review by cjnolet — APPROVED** [review]

File: cpp/include/raft/distance/detail/fused_distance_nn/custom_epilogue_with_broadcast.h
Comment (mdoijade) [reply cpp/include/raft/distance/detail/fused_distance_nn/custom_epilogue_with_broadcast.h]: done.

**Pull request merged**
